doc/tutorial/Server.lhs

```diff
@@ -1220,6 +1220,112 @@ $ curl http://localhost:8081/b
+
+Servant `0.5` introduced `Config` to handle this. The type machinery is beyond the scope of this tutorial, but the idea is simple: provide some data to the `serve` function, and that data is propagated to the functions that handle each combinator. Using `Config`, we can supply a function of type `BasicAuthData -> ExceptT ServantErr IO User` to the `BasicAuth` combinator handler. This will allow the handler to check authentication and return a `User` to downstream handlers if successful.
+
+In practice we wrap `BasicAuthData -> ExceptT ServantErr IO` into a slightly different function to better capture the semantics of basic authentication:
+
+``` haskell ignore
+-- | The result of authentication/authorization
+data BasicAuthResult usr
+  = Unauthorized
+  | BadPassword
+  | NoSuchUser
+  | Authorized usr
+  deriving (Eq, Show, Read, Generic, Typeable, Functor)
+
+-- | Datatype wrapping a function used to check authentication.
+newtype BasicAuthCheck usr = BasicAuthCheck
+  { unBasicAuthCheck :: BasicAuthData
+                     -> IO (BasicAuthResult usr)
+  }
+  deriving (Generic, Typeable, Functor)
+```
+
+Now, to use this within an API we demonstrate a complete example below. You will see the usage of `BasicAuthCheck` and `Config` in the `serverConfig` definition below!
+
+``` haskell
+-- | private data that needs protection
+newtype PrivateData = PrivateData { ssshhh :: Text }
+  deriving (Eq, Show, Generic)
+
+instance ToJSON PrivateData
+
+-- | public data that anyone can use.
+newtype PublicData = PublicData { somedata :: Text }
+  deriving (Eq, Show, Generic)
+
+instance ToJSON PublicData
+
+-- | A user we'll grab from the database when we authenticate someone
+newtype User = User { userName :: Text }
+  deriving (Eq, Show)
+
+-- | a type to wrap our public api
+type PublicAPI = Get '[JSON] [PublicData]
+
+-- | a type to wrap our private api
+type PrivateAPI = Get '[JSON] PrivateData
+
+-- | our API
+type API = "public"  :> PublicAPI
+      :<|> "private" :> BasicAuth "foo-realm" User :> PrivateAPI
+
+-- | a value holding a proxy of our API type
+api :: Proxy API
+api = Proxy
+
+-- | 'BasicAuthCheck' holds the handler we'll use to verify a username and password.
+authCheck :: BasicAuthCheck User
+authCheck =
+  let check (BasicAuthData username password) =
+        if username == "servant" && password == "server"
+        then return (Authorized (User "servant"))
+        else return Unauthorized
+  in BasicAuthCheck check
```

Comment:
I'm out-of-the-loop enough that this question may be answered somewhere else, but from this particular example I'm not sure how I would write an authorization algorithm that is predicated on any data in the request. I.e. Authorization rules will depend on the end-point but this function only works with `BasicAuthData`. So how could I write:

```haskell
type ViewEntity = "entity" :> Capture "id" String :> BasicAuth "my-app" User :> Get '[JSON] Entity
```

where authorization depends on *which* Entity `id` is being viewed. Authentication can often be answered "globally" (without regard to a particular request), but authorization often requires more information.